jaxquantum/core/measurements.py

```diff
@@ -40,133 +47,409 @@ def overlap(rho: Qarray, sigma: Qarray) -> Array:
+        # `compute_infidelity` is a "static" argument. JAX compiles a
+        # separate version of the code for each value of this flag.
+        if compute_infidelity:
+            rho = Qarray.create(_reconstruct_density_matrix(params, dim))
+            fid = fidelity(Qarray.create(true_rho_data), rho,
+                           force_positivity=True)
+            infidelity = 1.0 - fid
+        else:
+            infidelity = jnp.nan
+
+        new_carry = (new_params, new_opt_state)
+        history = {
+            "loss": loss_val,
+            "grads": grads,
+            "params": params,
+            "infidelity": infidelity,
+        }
+        return new_carry, history
+
+    initial_carry = (initial_params, initial_opt_state)
+    final_carry, history = lax.scan(
+        train_step, initial_carry, jnp.arange(epochs), length=epochs
+    )
+    return final_carry, history
+
+
+class MLETomographyResult(NamedTuple):
+    rho: Qarray
+    params_history: list
+    loss_history: list
+    grads_history: list
+    infidelity_history: Optional[list]
+
+
+class QuantumStateTomography:
+    def __init__(
+        self,
+        rho_guess: Qarray,
+        measurement_basis: Qarray,
+        measurement_results: jnp.ndarray,
+        complete_basis: Optional[Qarray] = None,
+        true_rho: Optional[Qarray] = None,
+    ):
+        """
+        Reconstruct a quantum state from measurement results using quantum state tomography.
+        The tomography can be performed either by direct inversion or by maximum likelihood estimation.
 
-    space_size = physical_basis.bdims[-1]
+        Args:
+            rho_guess (Qarray): The initial guess for the quantum state.
+            measurement_basis (Qarray): The basis in which measurements are performed.
+            measurement_results (jnp.ndarray): The results of the measurements.
+            complete_basis (Optional[Qarray]): The complete basis for state 
+            reconstruction used when using direct inversion. 
+            Defaults to the measurement basis if not provided.
+            true_rho (Optional[Qarray]): The true quantum state, if known.
+
+        """
+        self.rho_guess = rho_guess.data
+        self.measurement_basis = measurement_basis.data
+        self.measurement_results = measurement_results
+        self.complete_basis = (
+            complete_basis.data
+            if (complete_basis is not None)
+            else measurement_basis.data
+        )
+        self.true_rho = true_rho
+        self._result = None
 
-    for i in tqdm(range(space_size), total=space_size):
-        p_i = (rho @ physical_basis[i]).trace()
-        dm += p_i * logical_basis[i].data
+    @property
+    def result(self) -> Optional[MLETomographyResult]:
+        return self._result
 
-    return Qarray.create(dm, dims=logical_basis.dims, bdims=physical_basis[0].bdims)
+    def parameterize_density_matrix(self, params: jnp.ndarray, dim: int) -> jnp.ndarray:
```

Comment:
What's the point of this method if it just calls a function, with no use of `self`? Does this need to exist?